frontend/src/pages/CreatePipeline/index.tsx

```diff
@@ -0,0 +1,686 @@
+                }
+
+                return {
+                    ...node,
+                    data: {
+                        ...node.data,
+                        needs: needs,
+                    },
+                };
+            })
+        );
+    }
+
+
+    const setNodeDataIn = (affectedNode: Node, dataIn: FieldDescription[]) => {
+        setNodesRef.current((nds) =>
+            nds.map((node) => {
+                if (node.id !== affectedNode.id) {
+                    return node;
+                }
+
+                return {
+                    ...node,
+                    data: {
+                        ...node.data,
+                        dataIn: dataIn,
+                    },
+                };
+            })
+        );
+    }
+
+    const onAddEntryInput = (defaultName: string) => {
+        const entryNode = nodesRef.current.find((node) => node.type === "entryNodeEdit")
+        if (!entryNode) return;
+        const dataIn = [...entryNode.data.dataIn];
+        const inputField = new FieldDescription();
+        inputField.type = [];
+        inputField.name = defaultName;
+        dataIn.push(inputField);
+        setNodeDataIn(entryNode, dataIn);
+    }
+
+    const onDeleteEntryInput = (index: number) => {
+        const entryNode = nodesRef.current.find((node) => node.type === "entryNodeEdit")
+        if (!entryNode) return;
+        const dataIn = [...entryNode.data.dataIn];
+        dataIn.splice(index, 1)
+        setNodeDataIn(entryNode, dataIn)
+    }
+
+    const createEntryNode = () => {
+        const dataIn: FieldDescription[] = [];
+        const id = getId();
+        return {
+            id: `entry${id}`,
+            type: "entryNodeEdit",
+            data: {
+                onAddEntryInput: onAddEntryInput,
+                onSelectEntryInput: onSelectEntryInput,
+                onSelectEntryInputName: onSelectEntryInputName,
+                onDeleteEntryInput: onDeleteEntryInput,
+                identifier: "entry",
+                dataIn: dataIn,
+                label: "Pipeline entry",
+            },
+            position: {x: 50, y: 200},
+        }
+    }
+
+    const createExitNode = () => {
+        const dataOut: FieldDescription[] = [];
+        const id = getId();
+        return {
+            id: `exit${id}`,
+            type: "exitNodeEdit",
+            data: {
+                identifier: "exit",
+                dataOut: dataOut,
+                label: "Exit",
+            },
+            position: {x: 550, y: 200},
+        }
+    }
+
+    const addServiceNode = (serviceName: string, serviceId: string, tags: any[], serviceSlug: string, dataIn: FieldDescription[], dataOut: FieldDescription[]) => {
+        let counter = 2;
+        const selectedDataIn = new Array<string>(dataIn.length);
+        let identifier = serviceSlug
+        let label = serviceName;
+        while (nodesRef.current.find((node) => node.data.identifier === identifier)) {
```

Comment:
ESLint: Function declared in a loop contains unsafe references to variable(s) 'index'.(no-loop-func)